### DISCUSSION TITLE
Checksum errors when PDA is turned on

### DISCUSSION THREAD

**Discussion by keith-alphonso** [aqualinkd/AqualinkD#277] [Q&A]
I've got pretty much everything working on my (unfortunate) PDA-only board (PDA PS-8 combo).  The only problem I have is that I cannot use the original PDA.  As soon as I turn on the PDA I get a string of bad checksum errors, which then never stop even if I turn off the remote.  I'm thinking some thing to do with the RS485 wiring, but I'm not sure how to test or resolve.  I'm using a raspberry pi zero W with the RS485 adapter suggested in the wiki (FTDI USB-RS485-WE-1800-BT Cable, USB to RS485 Serial, 1.8M, Wire END).  I connected the data+ and data- wires only.  I tried also connecting the ground (black) wire to common, but that didn't help anything.
 

Any help appreciated!!

**Comment by sfeakes** [comment]
PDA only boards can only handle one device at a time unfortunately.  There is a setting in aqualinkd.conf for this. `pda_sleep_mode = yes`, make sure that’s set and AqualinkD will sleep when it’s not doing anything, and their for not create the problem of the two conflicting.
But you shouldn’t see any checksum errors, and if you do they should stop once the PDA is in sleep.  Can you explain the physical wiring you have, and al devices like SWG / JXi heater / VSP etc etc.
```
  Ie all in serial like 
control panel -> wired PDA remote -> RS485 adapter.

Or more like
Control panel |-> wired remote
              |-> RS485 adapter
```
Your description does sound like the PDA wire termination is conflicting some how with the RS485 termination.

**Reply by keith-alphonso** [reply]
My panel has two RS485 blocks, and I used the empty one....I suppose the other is for the remote.  So I didn't have to wire it on top of another, should be just on the bus.  sleep_mode is set to yes, and I made sure no browsers were open.  Output using the -d option looks like below, and continues even when remote is turned off.  

Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type  PDA HlightChars | HEX: 0x10|0x02|0x00|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type  PDA HlightChars | HEX: 0x10|0x02|0x00|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xe1' | HEX: 0x10|0x02|0x00|0xe1|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xf1' | HEX: 0x10|0x02|0x00|0xf1|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xff' | HEX: 0x10|0x02|0x00|0xff|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xff' | HEX: 0x10|0x02|0x00|0xff|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type  PDA HlightChars | HEX: 0x10|0x02|0x00|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type  PDA HlightChars | HEX: 0x10|0x02|0x00|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Ack w/ Command | HEX: 0x10|0x02|0x00|0x01|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xf9' | HEX: 0x10|0x02|0x00|0xf9|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xfd' | HEX: 0x10|0x02|0x00|0xfd|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xfd' | HEX: 0x10|0x02|0x00|0xfd|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xff' | HEX: 0x10|0x02|0x00|0xff|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xff' | HEX: 0x10|0x02|0x00|0xff|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x80 of type  PDA HlightChars | HEX: 0x10|0x02|0x80|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xe1' | HEX: 0x10|0x02|0x00|0xe1|0x10|0x02|0x81|0x00|0x93|0x10|0x03|
Warning: RS Serial: Serial read bad Jandy checksum, ignoring
Warning: RS Serial:  Read BAD PACKET To 0x00 of type   Unknown '0xf9' | HEX: 0x10|0x02|0x00|0xf9|0x10|0x02|0x50|0x11|0x19|0x8c|0x10|0x03|

Attached is a picture of the RS485 connection.

![IMG_4601](https://github.com/sfeakes/AqualinkD/assets/25887037/92ed8aef-a3fd-4549-ac9a-5f2db9334534)

**Comment by sfeakes** [comment]
Your connection looks good.  What version of AqualinkD are you using?

**Reply by sfeakes** [reply]
Have you got a Salt Water Generator connected to the RS485 bus?
Those messages a defiantly bad, just a guess, but it kinda looks like you have two devices sending a message at the same time.  Something is sending a message to the control panel and at the same time the control panel is sending a message to a SWG.  And those messages are colliding.

Can you post AqualinkD version and config file you are using. 
And can you post the output of the serial_logger, and make sure you use the Jandy PDA remote while the serial logger is running.

**Reply by keith-alphonso** [reply]
Aqualink Daemon v2.3.3

Output from serial logger with remote on:

Notice:  RS Serial: Suggested aqualinkd.conf values
Notice:  RS Serial: -------------------------
Notice:  RS Serial: panel_type =  PDA-PS8 Combo
Notice:  RS Serial: extended_device_id = 0x33
Notice:  RS Serial: -------------------------

Now if I run without the remote turned on:

Notice:  RS Serial: Suggested aqualinkd.conf values
Notice:  RS Serial: -------------------------
Notice:  RS Serial: panel_type =  PDA-PS8 Combo
Notice:  RS Serial: device_id = 0x60
Notice:  RS Serial: extended_device_id = 0x33
Notice:  RS Serial: -------------------------

[aqualinkd.conf.txt](https://github.com/sfeakes/AqualinkD/files/15066776/aqualinkd.conf.txt)

**Reply by keith-alphonso** [reply]
Also, I do have an SWG installed.

**Comment by sfeakes** [comment]
I’m sorry can you post the full output from serial_logger, (maybe run it standalone if you aren’t already).  I’m looking to see if you get any checksum errors with just using the serial_logger, and also looking to see if serial_logger can pull the panel info.
So something like the below command.
```sudo serial_logger -p 1000 /dev/ttyUSB0```

Nothing to do with your current problem, but what’s very interesting here is it looks like your panel can also communicate on 0x33 which is a OneTouch ID.  I’ve never seen any PDA only panel that can use anything other than 0x60, the PDA id.    And I see a bug in serial_logger due to this, the recommended settings it’s telling you are wrong.  Extended_device_id is invalid for PDA only panels at the moment, but if that truely is available, I could make PDA only panels a far more stable in AqualinkD

**Reply by keith-alphonso** [reply]
AqualinkD serial_logger V2.0
Notice:  RS Serial: Port /dev/ttyUSB0 low latency mode is set
Notice:  RS Serial: Logging serial information!
Please wait.|

Notice:  RS Serial: RS485 interface received 1000 packets in 30 seconds (~33.33 Msg/Sec)
Notice:  RS Serial: Jandy Control Panel Model   :  PDA-PS8 Combo
Notice:  RS Serial: Jandy Control Panel Version :       PDA: 7.1.0
Notice:  RS Serial: Jandy ID's found
Notice:  RS Serial: ID 0x33 is not used  <-- can use for Aqualinkd (Prefered Extended Device ID)
Notice:  RS Serial: ID 0x50 is in use    <-- Salt Water Generator (Aquarite mode)
Notice:  RS Serial: ID 0x60 is not used  <-- can use for Aqualinkd (PDA mode only)
Notice:  RS Serial: 

Notice:  RS Serial: Suggested aqualinkd.conf values
Notice:  RS Serial: -------------------------
Notice:  RS Serial: panel_type =  PDA-PS8 Combo
Notice:  RS Serial: device_id = 0x60
Notice:  RS Serial: extended_device_id = 0x33
Notice:  RS Serial: -------------------------

**Reply by keith-alphonso** [reply]
This is with the remote on:

Notice:  RS Serial: RS485 interface received 1000 packets in 24 seconds (~41.67 Msg/Sec)
Notice:  RS Serial: Jandy Control Panel Model   :  PDA-PS8 Combo
Notice:  RS Serial: Jandy Control Panel Version :       PDA: 7.1.0
Notice:  RS Serial: Jandy ID's found
Notice:  RS Serial: ID 0x60 is in use    <-- PDA Remote
Notice:  RS Serial: ID 0x80 is not used 
Notice:  RS Serial: ID 0x50 is in use    <-- Salt Water Generator (Aquarite mode)
Notice:  RS Serial: ID 0x33 is not used  <-- can use for Aqualinkd (Prefered Extended Device ID)
Notice:  RS Serial: ID 0x38 is not used 
Notice:  RS Serial: ID 0x58 is not used 
Notice:  RS Serial: ID 0x68 is not used 
Notice:  RS Serial: ID 0x81 is not used 
Notice:  RS Serial: 

Notice:  RS Serial: Suggested aqualinkd.conf values
Notice:  RS Serial: -------------------------
Notice:  RS Serial: panel_type =  PDA-PS8 Combo
Notice:  RS Serial: extended_device_id = 0x33
Notice:  RS Serial: -------------------------

**Reply by sfeakes** [reply]
Thanks for all the details, I'm not seeing anything out of the ordinary here.

I'm not sure this is a wiring issue, but since it's the only device on that block. Try Terminate the RS485, you can do this by connecting Brown / Green wires together on the end of the adapter.

If that doesn't improve, try changing the timing of the RS485 connection in aqualinkd.conf change `rs485_frame_delay = 4`, set it to fast as possible 0, if that doesn't work, slow it down to 8 or 12.

**Comment by ballle98** [comment]
@keith-alphonso I used to have an older PDA-4 combo board (FW version 1.2).  With that controler I could use a PDA and aqualinkd at the same time.  It would pobe for multiple PDAs (0x60, 0x61,0x62, 0x63).  So I configured aqualinkd at 0x62 and left PDA configured to 0x60.  When aqualinkd 0x62 was active it would not probe 0x60 but when it went to sleep (stopped acking messages from control panel) the control panel  would probe for all the PDA device IDs and the PDA could start a session.  Basically the RS485 is a party line where everyone sees the messages for everyone else you ignore the messages that are not for your device ID.  I just had to replace my control panel with a newer zodiac board PDA-PS6 Combo FW ver 7.1.0.  I noticed this controller only probes one PDA ID 0x60 so I had to change my aqualinkd config to use this ID.  I don't think you can have aqualinkd running and use a PDA with this board because I'm not aware of a state machine in aqualinkd that knows if to ignore messages sent to it's ID if it wakes up and has not seen a probe since it woke up.

**Reply by keith-alphonso** [reply]
Ok, so same problem here.  I don't need the remote very often, but when I do need it I want to make it fairly easy.  I created a workaround (hack).  I used xinetd to create 2 services, one to stop aqualinkd and one to start it, and run each on a different port.  The stop service stops aqualinkd and outputs a small html file that provides a link to the start service.  The start service starts the aqualinkd and outputs a small html file that redirects the user to the original page.  I then edited controller to add a small button that calls the stop service.  So I now have a button on my phone remote to enable the PDA remote for when I need to use it.  Problem solved.  

A far more elegant solution would be to add a button to disable the RS485 processing, but leave everything else running, then a button to re-enable it.  I just don't have time to figure out how to do that at the moment.

**Reply by ballle98** [reply]
for reference I found where I originally got this working in 2019 #40 but this was with a control panel supporting multiple PDA ids.  

The fix for this issue on my fork will get you closer to working - https://github.com/ballle98/AqualinkD/issues/26

need to rework the pda_shouldSleep() function so that when it's about to wake up it sets _pda_first_probe_recvd to false when loop count is set to 0 and return true.  might be able to do a dev branch that you could try.  

I didn't hook up my PDA when I replaced the panel.  I use https://github.com/earlephilhower/aquaweb when I need to configure the panel.  After I got aqualinkd connected to openhab and then to alexa and google no one ever used the pda it was faster to say "alexa turn on filter pump".  I see there is new simulor pages added to aqualinkd but they don't work on PDA panels yet :(

**Comment by ballle98** [comment]
Try this dev branch.  Make sure not to have a web browser open to aqualinkd as that will keep it awake.  Also make sure "pda_sleep_mode = yes" is set in /etc/aqualinkd.conf.  When aqualinkd is sleeping for 30 sec between refreshes if you turn your physical PDA on aqualinkd should stay asleep until the physical PDA is off.

```
sudo apt-get install git libsystemd-dev
cd ~
mkdir git2
cd git2
git clone https://github.com/ballle98/AqualinkD.git
git checkout dev/share_pda_id
make clean
make

```

You can test it like so
```
sudo systemctl stop aqualinkd
sudo release/aqualinkd -v -d -c /etc/aqualinkd.conf

```

If it works 
`sudo make install
`

otherwise
`sudo systemctl stop aqualinkd
`

**Reply by keith-alphonso** [reply]
What I have now will probably work for me, as you say since i have the phone remote I should not need the PDA very often.  Once I get the scheduler setup on aqualinkd the only thing I would need the PDA for is to change the color of the lights.

One other wish that I have is to show the messages sent by the system to the PDA, as this can clear some confusion that occurs.  for example, trying to turn off the filter pump while the heater is on will return "Pump to remain on during cooldown" and when switching to spa mode the message "Pump will turn on after delay".  Not having these messages can confuse the user as they think it is non-responsive.

**Reply by locnho** [reply]
I had tried this and ballle98 branch dev/share_pda_id does work. But I would say that it is 100% reliable. What I did instead:
1. Using the latest AquaLinkD branch
2. Add "BOOST POOL" for freeze protection
3. Add in ballle fix for fix "https://github.com/aqualinkd/AqualinkD/issues/78: PDA Init :- wait for next menu failed"

In the above and setup the button correct, I was able to get it working. Had this integrated with MQTT and HomeBridge. Hasn't add to HomeKit yet as I still have the LED to check out.

As to make this works with PDA, here what I did:
1. Setup a HomeBridge Cmd2 to disable or enable AquaLinkD service
2. Now, if I want to use the PDA, just disable AquaLinkD from HomeKit. This will avoid having AquaLinkD and PDA on at the same time.

If anyone want to have this for this PDA-PS6 Combo panel, here is the GIT - https://github.com/locnho/AqualinkD.git.

Thanks Ballle98 for the work.
